docs/proposals/20210524-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,564 @@
+          certFile: "/etc/kubernetes/pki/apiserver-etcd-client.crt"
+          keyFile: "/etc/kubernetes/pki/apiserver-etcd-client.key"
+    initConfiguration:
+      nodeRegistration:
+        criSocket: /var/run/containerd/containerd.sock
+        kubeletExtraArgs: {eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'}
+    joinConfiguration:
+      nodeRegistration:
+        criSocket: /var/run/containerd/containerd.sock
+        kubeletExtraArgs: {eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'}
+  version: "v1.20.2"
+```
+
+![etcdcluster-sequence](images/managed-etcd/etcdcluster-sequence.png)
+
+- Cluster.Spec will have a new field `ManagedExternalEtcdRef` of type `ObjectReference` (same as `ControlPlaneRef`). For using CAPI managed etcd cluster, user will set this field in Cluster manifest as shown above.
+- Cluster controller's `reconcileControlPlane` will check if `managedExternalEtcdRef` is set on cluster spec and if External etcd cluster is not `Ready`, it will pause the control plane provisioning by setting clusterv1.Paused annotation (`cluster.x-k8s.io/paused`) on it. This will allow etcd cluster provisioning to occur first.
+- Etcd cluster controller will generate a CA cert-key pair to be used for all etcd members. This external etcd CA will be saved in a Secret with name `{cluster.Name}-managed-etcd`. We can use the existing util/secret pkg in CAPI for this and add a new [Purpose](https://github.com/mrajashree/cluster-api/blob/etcdadm_bootstrap/util/secret/consts.go#L50) for managed etcd.
+- The external etcd controller will also save the etcd CA cert, and the apiserver client cert-key pair in Secrets named `{cluster.Name}-etcd` and `{cluster.Name}-apiserver-etcd-client` respectively.
+- Then the controller will first initialize etcd cluster by creating one EtcdadmConfig and a corresponding Machine resource. 
+- The CAPI [machine controller](https://github.com/mrajashree/cluster-api/blob/etcdadm_bootstrap/controllers/machine_controller_phases.go#L322-L378) gets the machine's IP address/hostname from the InfraMachine, and as soon as the first etcd machine's IP is obtained, it will create a Secret to store this address, and set the `Initialized` condition on the EtcdCluster object. This step is required because etcdadm join command must have the first node's client URL.
+- Once the EtcdCluster is Initialized, the etcd cluster controller will scale up the cluster to add as many members as required to match EtcdCluster.Spec.Replicas field
+- At the end of every reconciliation loop, the etcd cluster controller will check if desired number of members are created by
+    * Getting Machines owned by EtcdCluster
+    * Running healthcheck for each Machine with the etcd `/health` endpoint at `https://client-url:2379/health`
+- Once all members are ready, EtcdCluster will set the endpoints on EtcdCluster.Status.Endpoints field, and set EtcdCluster.Status.Ready to true
```

Comment:
I know this may be out of scope. How are `EtcdCluster.Status.Endpoints` consumed by kube-apiserver?

Reply:
good question, not out of scope, the section `Contract between etcd provider and control plane provider` talks about how control plane providers can consume this field. After that, each control plane provider is responsible for processing the endpoints field and passing it to the kube-apiserver. The default control plane provider in CAPI is KubeadmControlPlane (KCP), and it already has a field for accepting external etcd endpoints and passing them to kube-apiserver. So this default control plane provider will continue using the same logic